staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,619 @@
+type queueSetImpl struct {
+	lock                 sync.Mutex
+	name                 string
+	counter              fq.GoRoutineCounter
+	queues               []*fq.Queue
+	clk                  clock.PassiveClock
+	vt                   float64
+	estimatedServiceTime float64
+	lastRealTime         time.Time
+	robinIdx             int
+	// numRequestsEnqueued is the number of requests currently enqueued
+	// (eg: incremeneted on Enqueue, decremented on Dequue)
+	numRequestsEnqueued int
+	concurrencyLimit    int
+	desiredNumQueues    int
+	queueLengthLimit    int
+	requestWaitLimit    time.Duration
+	emptyHandler        fq.EmptyHandler
+	dealer              *shufflesharding.Dealer
+}
+
+// initQueues is a helper method for initializing an array of n queues
+func initQueues(n, baseIndex int) []*fq.Queue {
+	fqqueues := make([]*fq.Queue, n, n)
+	for i := 0; i < n; i++ {
+		fqqueues[i] = &fq.Queue{Index: baseIndex + i, Requests: make([]*fq.Request, 0)}
+
+	}
+	return fqqueues
+}
+
+// newQueueSetImpl creates a new queueSetImpl from passed in parameters
+func newQueueSetImpl(name string, concurrencyLimit, desiredNumQueues, queueLengthLimit, handSize int,
+	requestWaitLimit time.Duration, clk clock.PassiveClock, counter fq.GoRoutineCounter) (*queueSetImpl, error) {
+	dealer, err := shufflesharding.NewDealer(desiredNumQueues, handSize)
+	if err != nil {
+		return nil, errors.Wrap(err, "shuffle sharding dealer creation failed")
+	}
+
+	fq := &queueSetImpl{
+		counter:          counter,
+		name:             name,
+		queues:           initQueues(desiredNumQueues, 0),
+		clk:              clk,
+		vt:               0,
+		lastRealTime:     clk.Now(),
+		desiredNumQueues: desiredNumQueues,
+		concurrencyLimit: concurrencyLimit,
+		queueLengthLimit: queueLengthLimit,
+		requestWaitLimit: requestWaitLimit,
+		dealer:           dealer,
+	}
+	return fq, nil
+}
+
+func (qs *queueSetImpl) lockAndSyncTime() {
+	qs.lock.Lock()
+	qs.synctime()
+}
+
+// SetConfiguration is used to set the configuration for a queueSetImpl
+// update handling for when fields are updated is handled here as well -
+// eg: if desiredNumQueues is increased, SetConfiguration reconciles by
+// adding more queues.
+func (qs *queueSetImpl) SetConfiguration(concurrencyLimit, desiredNumQueues,
+	queueLengthLimit, handSize int, requestWaitLimit time.Duration) error {
+	// lock required as method can change Queues which has its indexes and length used
+	// concurrently
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+
+	// Adding queues is the only thing that requires immediate action
+	// Removing queues is handled by omitting indexes >desiredNumQueues from
+	// chooseQueueIdx
+
+	numQueues := len(qs.queues)
+	if desiredNumQueues > numQueues {
+		qs.queues = append(qs.queues,
+			initQueues(desiredNumQueues-numQueues, len(qs.queues))...)
+	}
+
+	dealer, err := shufflesharding.NewDealer(desiredNumQueues, handSize)
+	if err != nil {
```

Comment:
Do this before any side-effects on `qs`.

Reply:
Changed to be first operation after lock.